### ISSUE TITLE
Function name should not be changed when creating "AWS::Lambda::Function" resource type

### ISSUE THREAD

**Issue by JordanReiter** [spulec/moto#3144]
**how to reproduce the issue**

```
import json
import boto3
from moto import mock_iam, mock_cloudformation, mock_lambda


@mock_iam
@mock_cloudformation
@mock_lambda
def test_create_lambda_stack():
    assume_role_policy_document = {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Principal": {
                    "Service": ['lambda.amazonaws.com']
                },
                "Action": "sts:AssumeRole"
            }
        ]
    }
    iam = boto3.client('iam')
    create_role_resp = iam.create_role(
        RoleName='RoleName',
        AssumeRolePolicyDocument=json.dumps(assume_role_policy_document)
    )
    role_arn = create_role_resp['Role']['Arn']
    template = {
        "AWSTemplateFormatVersion": "2010-09-09",
        "Resources": {
            "CustomNameLambdaFunction": {
                "Type": "AWS::Lambda::Function",
                    "Properties": {
                        "Code": {
                            "ZipFile": (
                                "def hello(event, context):\n"
                                "    return { \n"
                                "        'message' : 'Hello'\n"
                                "    }\n"
                            )
                        },
                        "FunctionName": 'CustomName',
                        "Handler": 'handler.hello',
                        "Role": role_arn,
                        "Runtime": "python3.6",
                        "Timeout": 30,
                        "Environment": {
                            "Variables": {}
                        }
                    }
                }
        },
        "Outputs": {}
    }
    cf = boto3.client('cloudformation')
    cf.create_stack(StackName='StackName', TemplateBody=json.dumps(template))
    lmb = boto3.client('lambda')
    func = lmb.list_functions()['Functions'][0]
    assert func['FunctionName'] == 'CustomName', 'Function name was changed to ' + func['FunctionName']
    print("Created function", func['FunctionName'])


test_create_lambda_stack()
``` 
** What I expected to happen **
CloudFormation creates a function called `CustomName`

** What actually happens **
CloudFormation creates a function called ` StackName-CustomNameLambdaFunction-LDRA6MD8ZSE1` (last part is random string generated by `random_suffix` function).

Fix is straightforward: when the property `FunctionName` is provided, moto should have CloudFormation create the function using that name.

This is moto version **1.3.14**

The code to change is on `moto/cloudformation/parsing.py` under `NAME_TYPE_MAP`. You just need to add the line

        "AWS::Lambda::Function": "FunctionName",

to the map.

This matches up with the link immediately above NAME_TYPE_MAP, which states it is using https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-name.html as a reference.

According to that page, `AWS::Lambda::Function` is one of the supported resources for custom names, and the field in question is `FunctionName`.

**Comment by JordanReiter** [comment]
Actually looks like this is fixed in 1.3.15dev. Frustrating that this feature is not working on the released version, as it has been supported for a very long time.

**Comment by bblommers** [comment]
Hi @JordanReiter, we do have pre-releases for every PR that is merged: https://pypi.org/project/moto/#history, so the latest version should definitely support this.
